File: buildproj/MyDiskInit/TestMyFileSystem/MyBuf.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MyDiskInit.TestMyFileSystem
{
    class Buf
    {
        public int blkno;
        public char[] buffer;
        public Buf nextBuf;
        public Buf()
        {
            buffer = new char[512];
            nextBuf = null;
        }
    }

    class BufferManager
    {

	    /* static const member */
	    static const int NBUF = 15;			/* 缓存控制块、缓冲区的数量 */
	    static const int BUFFER_SIZE = 512;	/* 缓冲区大小。 以字节为单位 */
        Buf bFreeList;						/* 自由缓存队列控制块 */

        BufferManager()
        {
            bFreeList = new Buf();
            Buf temp = bFreeList;
            for(int i = 0; i < BUFFER_SIZE; i++)
            {
                temp.nextBuf = new Buf();
                temp = temp.nextBuf;
            }
        }

        /* 申请一块缓存，用于读写设备dev上的字符块blkno。*/
        public Buf GetBlk(int blkno)
        {
            Buf temp = bFreeList;
            while(temp.nextBuf != null)
            {
                if (temp.nextBuf.blkno == blkno)
                    break;
            }
            if (temp.nextBuf == null)
                temp = bFreeList;
            Buf reBuf = temp.nextBuf;
            temp.nextBuf = reBuf.nextBuf;
            return reBuf;
        }

        /* 释放缓存控制块buf */
	    public void Brelse(Buf bp);
        {
            bp.nextBuf = null;
        }		
	    public Buf Bread(int blkno);				/* 读一个磁盘块。blkno为目标磁盘块逻辑块号。 */
	    public void Bwrite(Buf bp);				/* 写一个磁盘块 */

    }
}

File: v6++proj/testMyFileSystem/MyBuf.cpp
#include "MyBuf.h"
#include "TestMyFileSystem.h"
#include <string>
#include <iostream>
using namespace std;
const std::string DISK_PATH = "C:\\Users\\saika\\Documents\\GitHub\\MyFileSystem\\buildproj\\MyDiskInit\\bin\\Debug\\MyDisk.img"; // 磁盘文件路径


Buf::Buf()
{
	blkno = 0;
	memset(buffer, 0, 512);
	nextBuf = NULL;
}

BufferManager::BufferManager()
{
	Buf *temp = &bFreeList;
	for (int i = 0; i < NBUF; i++)
	{
		temp->nextBuf = new Buf();
		temp = temp->nextBuf;
	}
}

BufferManager::~BufferManager()
{
	Buf *temp = bFreeList.nextBuf;
	Buf *t;
	while (temp)
	{
		t = temp;
		temp = temp->nextBuf;
		delete(t);
	}
}

Buf* BufferManager::GetBlk(int blkno)
{
	Buf* temp = &bFreeList;
	while (temp->nextBuf != NULL)
	{
		if (temp->nextBuf->blkno == blkno)
			break;
		temp = temp->nextBuf;
	}
	if (temp->nextBuf == NULL)
		temp = &bFreeList;
	Buf* reBuf = temp->nextBuf;
	temp->nextBuf = reBuf->nextBuf;
	return reBuf;
}

void BufferManager::Brelse(Buf* bp)
{
	bp->nextBuf = NULL;
	Buf* temp = &bFreeList;
	while (temp->nextBuf != NULL)
		temp = temp->nextBuf;
	temp->nextBuf = bp;
}


Buf* BufferManager::Bread(int blkno)
{
	Buf* reBuf = GetBlk(blkno);
	if (reBuf->blkno == blkno)
		return reBuf;
	reBuf->blkno = blkno;
	int fd = _open(DISK_PATH.c_str(), O_RDWR);
	_lseek(fd, blkno * 512, SEEK_SET);
	_read(fd, &reBuf->buffer, 512);
	_close(fd);
	return reBuf;
}

void BufferManager::Bwrite(Buf* bp)
{
	int fd = _open(DISK_PATH.c_str(), O_RDWR);
	_lseek(fd, bp->blkno * 512, SEEK_SET);
	_write(fd, bp->buffer, 512);
	_close(fd);
}

void BufferManager::ClrBuf(Buf *bp)
{
	/* 将缓冲区中数据清零 */
	for (unsigned int i = 0; i < BufferManager::BUFFER_SIZE; i++)
	{
		bp->buffer[i] = 0;
	}
	return;
}